### PR TITLE
Convert instances of ActionController::Parameters to Hashes before assigning default attributes

### LINKED ISSUE #69: default_value_for overrides values provided by an ActionController::Parameters object
My application uses permitted parameters when initializing a record. When attempting to provide nested attributes for a model that uses default_value_for, the ActionController::Parameters provided are ignored and overridden with the default value.

As of Rails 5, ActionController::Parameters no longer inherits from Hash. Because of this, the `initialize` method in `default_value_for.rb` does not handle instances of ActionController::Parameters correctly.

The issue lies at [default_value_for.rb#L127](https://github.com/FooBarWidget/default_value_for/blob/master/lib/default_value_for.rb#L127). If `attributes` is a hash, it sets `@initialization_attributes` to be the result of `attributes.stringify_keys`. Otherwise, it sets it to an empty Hash, `{}`

This should support instances of ActionController::Parameters as well as Hashes.

### PR DESCRIPTION
**Pull request by rymai**
This builds upon #70.

Convert initialization attributes to a hash if they respond to `to_h`.

1. This will allow to handle instances of `ActionController::Parameters` without making `ActionDispatch` a runtime dependency of this gem, as suggested by https://github.com/FooBarWidget/default_value_for/pull/70#pullrequestreview-180922504.
1. This also adds dedicated tests for the handling of `ActionController::Parameters`.
1. This also fixes an edge-case in tests when using Arel 5 (i.e. Rails <= 4.1) with Ruby >= 2.4.

Closes #69.

### Test results without the fix

```
› bundle exec appraisal rake test
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_32.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 3.2.22.5

Testing with Action Pack version 3.2.22.5

Run options: --seed 56411

# Running:

...................................

Finished in 0.291835s, 119.9308 runs/s, 181.6095 assertions/s.

35 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_40.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.0.13

Testing with Action Pack version 4.0.13

Run options: --seed 19827

# Running tests:

.....................................

Finished tests in 0.270046s, 137.0137 tests/s, 196.2629 assertions/s.

37 tests, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_41.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.1.16

Testing with Action Pack version 4.1.16

Run options: --seed 24243

# Running:

.....................................

Finished in 0.240286s, 153.9832 runs/s, 220.5705 assertions/s.

37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_42.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.2.10

Testing with Action Pack version 4.2.10

Run options: --seed 59744

# Running:

.....................................

Finished in 0.279463s, 132.3968 runs/s, 189.6494 assertions/s.

37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_50.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.0.7.1

Testing with Action Pack version 5.0.7.1

Run options: --seed 62880

# Running:

..............F

Failure:
DefaultValuePluginTest#test_doesnt_overwrite_explicitly_provided_nil_values_in_mass_assignment_with_action_controller_parameters [test.rb:411]:
Expected 1234 to be nil.


Traceback (most recent call last):
	17: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:63:in `block in autorun'
	16: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:136:in `run'
	15: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:159:in `__run'
	14: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:159:in `map'
	13: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:159:in `block in __run'
	12: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:319:in `run'
	11: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:347:in `with_info_handler'
	10: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:360:in `on_signal'
	 9: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:320:in `block in run'
	 8: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:320:in `each'
	 7: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:321:in `block (2 levels) in run'
	 6: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:334:in `run_one_method'
	 5: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:802:in `record'
	 4: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:802:in `each'
	 3: from /Users/remy/.gem/ruby/2.5.3/gems/minitest-5.11.3/lib/minitest.rb:803:in `block in record'
	 2: from /Users/remy/.gem/ruby/2.5.3/gems/railties-5.0.7.1/lib/rails/test_unit/reporter.rb:23:in `record'
	 1: from /Users/remy/.gem/ruby/2.5.3/gems/railties-5.0.7.1/lib/rails/test_unit/reporter.rb:70:in `format_rerun_snippet'
/Users/remy/.gem/ruby/2.5.3/gems/railties-5.0.7.1/lib/rails/test_unit/reporter.rb:70:in `method': undefined method `test_doesnt_overwrite_explicitly_provided_nil_values_in_mass_assignment_with_action_controller_parameters' for class `Minitest::Result' (NameError)
rake aborted!
Command failed with status (1): [/Users/remy/.rubies/ruby-2.5.3/bin/ruby te...]
/Users/remy/Code/GitLab/default_value_for/Rakefile:5:in `block in <top (required)>'
/Users/remy/.gem/ruby/2.5.3/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `load'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

```
› bundle exec appraisal rails-51 rake test
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_51.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.1.6.1

Testing with Action Pack version 5.1.6.1

Run options: --seed 57276

# Running:

........F

Failure:
DefaultValuePluginTest#test_doesnt_overwrite_explicitly_provided_nil_values_in_mass_assignment_with_action_controller_parameters [test.rb:411]:
Expected 1234 to be nil.


bin/rails test test.rb:408

............................

Finished in 0.391597s, 94.4849 runs/s, 135.3432 assertions/s.
37 runs, 53 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
Command failed with status (1): [/Users/remy/.rubies/ruby-2.5.3/bin/ruby te...]
/Users/remy/Code/GitLab/default_value_for/Rakefile:5:in `block in <top (required)>'
/Users/remy/.gem/ruby/2.5.3/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `load'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

```
› bundle exec appraisal rails-52 rake test
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_52.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.2.2

Testing with Action Pack version 5.2.2

Run options: --seed 28105

# Running:

...................................F

Failure:
DefaultValuePluginTest#test_doesnt_overwrite_explicitly_provided_nil_values_in_mass_assignment_with_action_controller_parameters [test.rb:411]:
Expected 1234 to be nil.


bin/rails test test.rb:408

.

Finished in 0.374031s, 98.9223 runs/s, 141.6995 assertions/s.
37 runs, 53 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
Command failed with status (1): [/Users/remy/.rubies/ruby-2.5.3/bin/ruby te...]
/Users/remy/Code/GitLab/default_value_for/Rakefile:5:in `block in <top (required)>'
/Users/remy/.gem/ruby/2.5.3/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `load'
/Users/remy/.gem/ruby/2.5.3/bin/bundle:23:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

### Test results with the fix

```
› bundle exec appraisal rake test
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_32.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 3.2.22.5

Testing with Action Pack version 3.2.22.5

Run options: --seed 7448

# Running:

...................................

Finished in 0.278227s, 125.7966 runs/s, 190.4919 assertions/s.

35 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_40.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.0.13

Testing with Action Pack version 4.0.13

Run options: --seed 61873

# Running tests:

.....................................

Finished tests in 0.275222s, 134.4369 tests/s, 192.5718 assertions/s.

37 tests, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_41.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.1.16

Testing with Action Pack version 4.1.16

Run options: --seed 40983

# Running:

.....................................

Finished in 0.286689s, 129.0597 runs/s, 184.8693 assertions/s.

37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_42.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 4.2.10

Testing with Action Pack version 4.2.10

Run options: --seed 17124

# Running:

.....................................

Finished in 0.235450s, 157.1459 runs/s, 225.1009 assertions/s.

37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_50.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.0.7.1

Testing with Action Pack version 5.0.7.1

Run options: --seed 37038

# Running:

.....................................

Finished in 0.321052s, 115.2461 runs/s, 165.0823 assertions/s.
37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_51.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.1.6.1

Testing with Action Pack version 5.1.6.1

Run options: --seed 40558

# Running:

.....................................

Finished in 0.408612s, 90.5504 runs/s, 129.7074 assertions/s.
37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
>> BUNDLE_GEMFILE=/Users/remy/Code/GitLab/default_value_for/gemfiles/rails_52.gemfile bundle exec rake test
/Users/remy/.rubies/ruby-2.5.3/bin/ruby test.rb

Testing with Active Record version 5.2.2

Testing with Action Pack version 5.2.2

Run options: --seed 6828

# Running:

.....................................

Finished in 0.391656s, 94.4707 runs/s, 135.3228 assertions/s.
37 runs, 53 assertions, 0 failures, 0 errors, 0 skips
```

### CODE REVIEW

**Comment by rymai** [comment]
@chrisarcand Could you please have a look? Thanks! ❤️

**Comment by rymai** [comment]
@chrisarcand Any chance to get this reviewed? :)

**Comment by chrisarcand** [comment]
Yup! This is in my to-dos and I'll definitely be able to get this done next week, along with some other bits of maintenance before the pending Rails 6 release next month.

Thank you for your patience.

File: test.rb [hunk 1]
```diff
@@ -22,11 +22,44 @@
...
 
 if ActiveSupport::VERSION::MAJOR == 3
   require 'active_support/core_ext/logger'
 end
 
+if ActionPack::VERSION::MAJOR > 3
+  require 'action_controller'
+end
+
+# Handle an edge-case when using Arel 5 (i.e. Rails <= 4.1) with Ruby >= 2.4:
```
Comment (chrisarcand) [inline test.rb]: Using Rails <= 4.1 with Ruby >= shouldn't really be supported, to be honest.

* Rails 5.0 was released in June 2016, thus ending all support (including severe security fixes) for Rails 4.1. 
* Ruby 2.4 was released _after_ that, in December.

With gems specifically used as plugins of Rails, I tend to follow those support guidelines as no one should assume a gem would guarantee support for a version spread like that when the framework itself definitely does not.

However, since you've already gone through a lot of in-depth trouble here to make this work, I'll certainly merge it. Please note if there are any issues with this patch it'll likely just get removed in the future, though.

Comment (rymai) [reply test.rb]: Sounds good! Yeah, I agree this edge-case shouldn't happen in reality and that's definitely ok to remove it in the future (I think I encountered this while testing the Rails/Ruby versions combinations manually).

**Comment by rymai** [comment]
> Thanks so much for this high quality pull request. Clear description, incorporating all feedback from the issue, well tested and even went out of your way for an edge case that I don't really think you needed to worry about.
> 
> This will be included in a release I'll make shortly with some other bits of things in anticipation of the Rails 6 release at the end of the month.

Thanks you! 🙏 🎉

**Comment by rymai** [comment]
@chrisarcand Hey, any chance to get a new release sometime? :)

**Comment by chrisarcand** [comment]
Hah, I suppose it's time to stop waiting for Rails 6 to actually drop ;)
